### DISCUSSION TITLE
SplitHTTP with or without Mux

### DISCUSSION THREAD

**Discussion by APT-ZERO** [XTLS/Xray-core#3582] [Q&A] [answered]
Is it advised to use SH-H1 SH-H2 SH-H3 with or without Mux?
also for other UDP based protocols like QUIC and KCP, is it better to enable Mux or not?
is it even effective on UDP based Protocols?

**Comment by mmmray** [comment, marked as answer]
for efficiency and stability: for h1 it definitely makes sense to enable mux. h1 can launch `1 * GET + maxConcurrentUploads * POST` requests (so, 11 requests) concurrently **per incoming connection**, because h1 as a protocol has no native mux, so either reduce `maxConcurrentUploads` and/or use mux to reduce the total connection amount. (or use h2/h3 which performs much better in general, but often that's not an option)

for anti-censorship: honestly it really depends on the firewall (and also, which firewall are we talking about), at this stage it's hard to tell if mux has the same camouflage effect on splithttp as it does on other transports. common wisdom is "more connections = bad", but at the same time, upload is separate from download...

for h2 and h3 it shouldn't matter in principle, but might depend on the CDN's performance and what its own overhead for a HTTP request is.

keep in mind that h1/h2/h3 is not only a choice between client and CDN, but also between CDN and origin server. If you use Cloudflare with h3 on the client, it still talks h1 or h2 to the origin server. Or if you use another CDN, it may only talk h1 to the origin server. In this case you can achieve better performance with mux, even if the client connection is h2 or h3.

**Reply by APT-ZERO** [reply]
Thank you, what do you think about SplitHTTP without TLS, is it possible to use it and remain undetected? (i guess no because of UUID in the path)
What if you make breaking changes in this transport, is it possible to make it remain undetectable without TLS?

**Reply by mmmray** [reply]
Ok time for a longer answer.

> is it possible to make it remain undetectable without TLS?

The answer is "no" for every CDN transport, websocket, httpupgrade, splithttp, it doesn't matter. Without TLS, all the transports are detectable in theory, and there hasn't been any focus **in any core** on making plaintext HTTP undetectable.

> (i guess no because of UUID in the path)

It's not just that. The entire idea of making undetectable plaintext HTTP connections is considered a completely lost cause in Xray and other cores. The GFW just needs to read a little bit past the headers, and then it sees either VMESS (random bytes) or VLESS (literally plaintext protocol that can be parsed). Even the user agent is still Go-http-client for this reason, because it's pointless and TLS is recommended anyway.

I guess some people use these transports without TLS anyway due to certain throttling behavior in Iran (and specifically Irancell what what I think is essentially a downgrade attack from TLS), but it's not sustainable.

I've seen this specifically with websocket. People use notls and then this happens: https://github.com/XTLS/Xray-core/discussions/3407

Of course, **it's understandable to use without TLS if it works**, but once the GFW updates, fixing it in core is either impossible or the fixes are:

1. many small fixes (annoying for maintainers, and potentially causing regressions in core, breaking compatibility)
2. easy for the GFW to follow up on
3. a bit arbitrary
4. potentially improving things in one country, but making things worse in another 
5. and also: Chinese core devs have a hard time understanding why this random header change is necessary for Iran, or if it is already obsolete and the code can be cleaned up again.

If you want to keep using notls in this situation, it's best to fork core for your country or, maybe, write a new transport that is actually properly designed around the notls requirement (still, it may only be useful for one country), or write some kind of dialerProxy-construct like fragment that allows you to "patch" the connection with random tweaks like this (although personally I don't have the creativity for such a thing)

**Comment by mmmray** [comment]
I forgot to answer some other parts of your questions

>  also for other UDP based protocols like QUIC and KCP, is it better to enable Mux or not? is it even effective on UDP based Protocols?

Cannot be said in a general way, in the case of splithttp+h3 (without CDN) and xray's QUIC transport, it is actually harmful because it applies head of line blocking where it is otherwise not necessary, but that's just making assumptions about how the protocol actually works.

~~This is besides the point, but if you construct MUX.cool with Wireguard somehow (dialerProxy?), it is also completely useless and probably doubly-harmful.~~
